File: src/shell_detect.nim
## Shell 检测模块
##
## 对应 C 源码：temp/src/tools.c `detect_shells()`
##
## 检测系统中可用的 shell，优先使用 $SHELL 环境变量（POSIX），
## 否则在 PATH 中搜索已知 shell 可执行文件。

import std/os

type
  ShellInfo* = object
    name*: string
    path*: string
    found*: bool

proc detectShells*(): seq[ShellInfo] =
  ## 检测系统中可用的 shell
  ##
  ## POSIX: 读取 $SHELL 环境变量
  ## Windows: 在 PATH 中搜索已知 shell 可执行文件
  ##
  ## 返回 seq[ShellInfo]，每个元素包含 shell 名称、路径和存在标记

  when not defined(windows):
    # POSIX 实现：读取 $SHELL 环境变量
    let shellEnv = getEnv("SHELL")
    if shellEnv.len > 0:
      let name = extractFilename(shellEnv)
      result.add(ShellInfo(
        name: name,
        path: shellEnv,
        found: true
      ))
  else:
    # Windows 实现：搜索已知 shell
    const shellNames = ["bash.exe", "pwsh.exe", "powershell.exe", "cmd.exe"]
    const extraBashPaths = [
      "C:\\Program Files\\Git\\bin\\bash.exe",
      "C:\\Program Files (x86)\\Git\\bin\\bash.exe",
      "C:\\msys64\\usr\\bin\\bash.exe",
      "C:\\mingw64\\usr\\bin\\bash.exe",
      "C:\\Git\\bin\\bash.exe"
    ]

    for name in shellNames:
      # 先在 PATH 中搜索
      let foundPath = findExe(name)
      if foundPath.len > 0:
        result.add(ShellInfo(
          name: name,
          path: foundPath,
          found: true
        ))
      elif name == "bash.exe":
        # bash 额外检查固定安装路径
        for extraPath in extraBashPaths:
          if fileExists(extraPath):
            result.add(ShellInfo(
              name: name,
              path: extraPath,
              found: true
            ))
            break

File: core/src/shell_detect.rs
#[derive(Debug, Clone)]
pub struct ShellInfo {
    pub name: String,
    pub path: String,
    pub found: bool,
}

pub fn detect_shells() -> Vec<ShellInfo> {
    let mut result = Vec::new();

    #[cfg(not(target_os = "windows"))]
    {
        if let Ok(shell_env) = std::env::var("SHELL")
            && !shell_env.is_empty()
        {
            let name = std::path::Path::new(&shell_env)
                .file_name()
                .map(|f| f.to_str().unwrap_or(""))
                .unwrap_or("")
                .to_string();
            result.push(ShellInfo {
                name,
                path: shell_env,
                found: true,
            });
        }
    }

    #[cfg(target_os = "windows")]
    {
        let shell_names = ["bash.exe", "pwsh.exe", "powershell.exe", "cmd.exe"];
        let extra_bash_paths = [
            r"C:\Program Files\Git\bin\bash.exe",
            r"C:\Program Files (x86)\Git\bin\bash.exe",
            r"C:\msys64\usr\bin\bash.exe",
            r"C:\mingw64\usr\bin\bash.exe",
            r"C:\Git\bin\bash.exe",
        ];

        for name in &shell_names {
            if let Ok(found_path) = which::which(name) {
                result.push(ShellInfo {
                    name: name.to_string(),
                    path: found_path.to_str().unwrap_or("").to_string(),
                    found: true,
                });
            } else if *name == "bash.exe" {
                for extra_path in &extra_bash_paths {
                    if std::path::Path::new(extra_path).exists() {
                        result.push(ShellInfo {
                            name: name.to_string(),
                            path: extra_path.to_string(),
                            found: true,
                        });
                        break;
                    }
                }
            }
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_shells_non_empty() {
        let shells = detect_shells();
        assert!(!shells.is_empty());
    }

    #[test]
    fn test_shell_name_not_empty() {
        let shells = detect_shells();
        for shell in &shells {
            assert!(!shell.name.is_empty());
        }
    }

    #[test]
    fn test_shell_path_not_empty() {
        let shells = detect_shells();
        for shell in &shells {
            assert!(!shell.path.is_empty());
        }
    }

    #[test]
    fn test_shell_found_flag() {
        let shells = detect_shells();
        for shell in &shells {
            assert!(shell.found);
        }
    }

    #[test]
    fn test_shell_path_exists() {
        let shells = detect_shells();
        for shell in &shells {
            assert!(std::path::Path::new(&shell.path).exists());
        }
    }

    #[test]
    fn test_common_shells() {
        let shells = detect_shells();
        let names: Vec<&str> = shells.iter().map(|s| s.name.as_str()).collect();
        let common = ["bash", "zsh", "sh", "fish"];
        let has_common = common.iter().any(|c| names.contains(c));
        assert!(has_common);
    }
}
